### PR TITLE
feat(*): support RPC style

### PR DESCRIPTION
**Pull request by iawia002**
**What this PR does / why we need it**:

Add a new builder to support serve RPC style APIs:

```
POST /?Action=GetUserByID&Version=2018-01-01
```

The default builder is still RESTful builder, here is an example to use RPC builder:

```go
var echo = definition.RPCDescriptor{
	Description: "Echo API",
	Middlewares: []definition.Middleware{reqlog.Default()},
	Consumes:    []string{definition.MIMEAll},
	Produces:    []string{definition.MIMEJSON},
	Actions: []definition.RPCAction{
		{
			Name:    "Echo",
			Version: "2020-10-10",
			Function: func(ctx context.Context, msg string) (string, error) {
				return msg, nil
			},
			Parameters: []definition.Parameter{
				{
					Source:      definition.Query,
					Name:        "msg",
					Description: "Corresponding to the second parameter",
				},
			},
			Results: definition.DataErrorResults(""),
		},
		{
			Name:    "Echo1",
			Version: "2020-10-10",
			Function: func(ctx context.Context, msg string) (string, error) {
				return msg, nil
			},
			Parameters: []definition.Parameter{
				{
					Source:      definition.Query,
					Name:        "msg",
					Description: "Corresponding to the second parameter",
				},
			},
			Results: definition.DataErrorResults(""),
		},
	},
}

func main() {
	cmd := config.NewDefaultNirvanaCommand()
	conf := nirvana.NewDefaultConfig()
	conf.Configure(
		nirvana.APIStyle(builder.APIStyleRPC),
		nirvana.Descriptor(echo, echo2),
	)
	if err := cmd.ExecuteWithConfig(conf); err != nil {
		log.Fatal(err)
	}
}
```

```
INFO  1119-18:25:59.051+08 config.go:309 | Listening on :8080
INFO  1119-18:25:59.052+08 builder.go:180 | Path: /?Version=2020-10-10&Action=Echo, Consumes: [*/*], Produces: [application/json]
INFO  1119-18:25:59.052+08 builder.go:180 | Path: /?Version=2020-10-10&Action=Echo1, Consumes: [*/*], Produces: [application/json]
```

And for generating API documentation for RPC style APIs, you need to add a `style` tag to the descriptors file:

```
// +nirvana:api=descriptors:"Descriptor" style:"rpc"
```

Screen Shot 2020-11-19 at 18 41 33

**Which issue(s) this PR is related to** *(optional, link to 3rd issue(s))*:

Fixes #

Reference to #

**Special notes for your reviewer**:

/cc @xpofei 

**Release note**:

```release-note
NONE
```

### CODE REVIEW

File: service/rpc/builder.go [hunk 1]
```diff
@@ -0,0 +1,235 @@
...
+	s := &rpcService{
+		executors: b.bindings,
+		filters:   b.filters,
+		logger:    b.logger,
+		producers: service.AllProducers(),
+	}
+	return s, nil
+}
+
+type rpcService struct {
```
Comment (xpofei) [inline service/rpc/builder.go:196]: 名字还是保持 service 吧，都已经在 rpc 包下了

File: service/rest/builder.go [hunk 2]
```diff
@@ -265,76 +246,52 @@ func (b *builder) Build() (Service, error) {
...
-//          ↓                    ↑
-// |-ParameterGenerators-|-DestinationHandlers-|
-//          ↓                    ↑
-// |------------User Function-----------|
-type Service interface {
-	http.Handler
-}
-
-type service struct {
+type restService struct {
```
Comment (xpofei) [inline service/rest/builder.go:266]: 还是保持叫 service ？

File: service/rpc/errors.go [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+limitations under the License.
+*/
+
+package rpc
+
+import (
+	"github.com/caicloud/nirvana/errors"
+)
+
+var noExecutorForPath = errors.MethodNotAllowed.Build("Nirvana:Service:NoExecutorForPath", "no executor for path ${path}")
```
Comment (xpofei) [inline service/rpc/errors.go:23]: noExecutorForURL 或者 ForAction 之类的比较好

File: service/types.go [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+	AddDescriptor(descriptors ...interface{}) error
+	// Middlewares returns all router middlewares.
+	Middlewares() map[string][]definition.Middleware
+	// Definitions returns all definitions. If a modifier exists, it will be executed.
+	Definitions() map[string][]definition.Definition
+	// Build builds a service to handle request.
+	Build() (Service, error)
+}
+
+// Service handles HTTP requests.
```
Comment (xpofei) [inline service/types.go:50]: 这里的流程注释迁移到 rest 的 service 下

File: definition/rpc.go [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package definition
+
+// RPCDescriptor describes a descriptor for API definition in RPC style.
+type RPCDescriptor struct {
```
Comment (xpofei) [inline definition/rpc.go:20]: // Prefix describes url path prefix for all RPCActions

File: definition/rpc.go [hunk 6]
```diff
@@ -0,0 +1,72 @@
...
+	// It will override parent descriptor's tags.
+	Tags []string
+	// Consumes indicates content types that current definitions
+	// and child definitions can consume.
+	// It will override parent descriptor's consumes.
+	Consumes []string
+	// Produces indicates content types that current definitions
+	// and child definitions can produce.
+	// It will override parent descriptor's produces.
+	Produces []string
```
Comment (xpofei) [inline definition/rpc.go:36]: // Actions contain actions in this descriptor. These actions will inherit the Middlewares, Tags, Consumes, Produces of the descriptor if values in the action are not specified

File: examples/api-basic/api/v2/app_test.go [hunk 7]
```diff
@@ -22,12 +22,13 @@ import (
 	"testing"
 
 	"github.com/caicloud/nirvana/examples/api-basic/application"
+	"github.com/caicloud/nirvana/service/builder"
```
Comment (xpofei) [inline examples/api-basic/api/v2/app_test.go:26]: 这里多了个空行

**Comment by iawia002** [comment]
@xpofei All comments addressed, PTAL

File: service/rest/router/regexp.go [hunk 8]
```diff
@@ -70,15 +72,15 @@ func (n *regexpNode) Match(ctx context.Context, c Container, path string) (Execu
 		return nil, RouterNotFound.Error()
```
Comment (xpofei) [inline service/rest/router/regexp.go:72]: utils.go 里面的 error 就不用导出包了

**Pull request merged**